File: src/commands/git/start.ts
import * as prompts from "@clack/prompts";
import { Command } from "commander";
import { getShellError } from "../../utils/errors";
import { getDefaultBranch } from "../../utils/git";

export const start = new Command("start")
	.description("Create a new branch from a fresh default branch")
	.argument("<name>", "Name for the new branch")
	.action(async (name: string) => {
		prompts.intro("git start");

		const s = prompts.spinner();
		try {
			const defaultBranch = await getDefaultBranch();

			s.start(`Switching to ${defaultBranch} and pulling...`);
			await Bun.$`git checkout ${defaultBranch}`.quiet();
			await Bun.$`git pull`.quiet();
			s.stop(`${defaultBranch} is up to date`);

			s.start(`Creating branch ${name}...`);
			await Bun.$`git checkout -b ${name}`.quiet();
			s.stop(`On new branch ${name}`);

			prompts.outro("Ready to go!");
		} catch (e: unknown) {
			s.stop("Failed");
			prompts.log.error(getShellError(e));
			process.exit(1);
		}
	});

File: src/commands/git/start.rs
use anyhow::Result;
use crate::utils::git::{get_default_branch, Shell};

fn run_cmd(args: &[&str]) -> Result<String> {
    let output = std::process::Command::new("git").args(args).output()?;
    if !output.status.success() {
        let stderr = String::from_utf8_lossy(&output.stderr);
        return Err(anyhow::anyhow!("git failed: {stderr}"));
    }
    Ok(String::from_utf8_lossy(&output.stdout).to_string())
}

pub fn execute_start(shell: &dyn Shell, name: &str) -> Result<()> {
    let default = get_default_branch(shell)?;

    println!("Switching to {default} and pulling...");
    run_cmd(&["checkout", &default])?;
    run_cmd(&["pull"])?;
    println!("{default} is up to date");

    println!("Creating branch {name}...");
    run_cmd(&["checkout", "-b", name])?;
    println!("On new branch {name}");

    println!("Ready to go!");
    Ok(())
}
